core/base/polymorphic_object.hpp

```diff
@@ -0,0 +1,521 @@
+ * `U` should implement ConvertibleTo<V>.
+ * Then, the implementation of `PolymorphicObject::copy_from` automatically
+ * generated by `EnablePolymorphicObject` mixin will use RTTI to figure out that
+ * `U` implements the interface and convert it using the convert_to / move_to
+ * methods of the interface.
+ *
+ * As an example, the following function:
+ *
+ * ```c++
+ * void my_function(const U *u, V *v) {
+ *     v->copy_from(u);
+ * }
+ * ```
+ *
+ * will convert object `u` to object `v` by checking that `u` can be dynamically
+ * casted to `ConvertibleTo\<V\>`, and calling
+ * ConvertibleTo\<V\>::convert_to(V*)` to do the actual conversion.
+ *
+ * In case `u` is passed as a unique_ptr, call to `convert_to` will be replaced
+ * by a call to `move_to` and trigger move semantics.
+ *
+ * @tparam ResultType  the type to which the implementer can be converted to,
+ *                     has to be a subclass of PolymorphicObject
+ */
+template <typename ResultType>
+class ConvertibleTo {
+public:
+    using result_type = ResultType;
+
+    virtual ~ConvertibleTo() = default;
+
+    /**
+     * Converts the implementer to an object of type result_type.
+     *
+     * @param result  the object used to store the result of the conversion
+     */
+    virtual void convert_to(result_type *result) const = 0;
+
+    /**
+     * Converts the implementer to an object of type result_type by moving data
+     * from this object.
+     *
+     * This method is used when the implementer is a temporary object, and move
+     * semantics can be used.
+     *
+     * @param result  the object used to emplace the result of the conversion
+     *
+     * @note ConvertibleTo::move_to can be implemented by simply
+     *       ConvertibleTo::convert_to. However, this operation can often be
+     *       optimized by exploiting the fact that implementer's data can be
+     *       moved to the result.
+     */
+    virtual void move_to(result_type *result) = 0;
+};
+
+
+/**
+ * This mixin inherits from (a subclass of) PolymorphicObject and provides a
+ * base implementation of a new concrete polymorphic object.
+ *
+ * The mixin changes parameter and return types of appropriate public methods of
+ * PolymorphicObject in the same way EnableAbstractPolymorphicObject does.
+ * In addition, it also provides default implementations of PolymorphicObject's
+ * vritual methods by using the _executor default constructor_ and the
+ * assignment operator of ConcreteObject. Consequently, the following is a
+ * minimal example of PolymorphicObject:
+ *
+ * ```c++
+ * struct MyObject : EnablePolymorphicObject<MyObject> {
+ *     MyObject(std::shared_ptr<const Executor> exec)
+ *         : EnablePolymorphicObject<MyObject>(std::move(exec))
+ *     {}
+ * };
+ * ```
+ *
+ * In a way, this mixin can be viewed as an extension of default
+ * constructor/destructor/assignment operators.
+ *
+ * @note  This mixin does not enable copying the polymorphic object to the
```

Comment:
Am I misunderstanding something here ? `EnablePolymorphicObject` has the `copy_from_impl(Polymorphic *other)` which copies the same types right ? And `EnablePolymorphicAssignment` class is copying two different types below ?